### PR TITLE
x-wing: add fallible `DecapsulationKeyRejectNonContrib`

### LINKED ISSUE #364: x-wing: Non-contributory behaviour needs to be controllable
The `x-wing` crate currently implements [`draft-connolly-cfrg-xwing-kem-06`](https://datatracker.ietf.org/doc/html/draft-connolly-cfrg-xwing-kem-06) (AFAICT there are no breaking changes between that and the current draft `-10`). That draft uses the X25519 function from [Section 5 of RFC 7748](https://www.rfc-editor.org/info/rfc7748/#section-5), and the standard basepoint from [Section 6.1](https://www.rfc-editor.org/info/rfc7748/#section-6.1). However, there is another part of Section 6.1 that affects Xwing:

> Both now share K = X25519(a, X25519(b, 9)) = X25519(b, X25519(a, 9)) as a shared secret.  Both MAY check, without leaking extra
> information about the value of K, whether K is the all-zero value and abort if so (see below).
> [...]
> The check for the all-zero value results from the fact that the X25519 function produces that value if it operates on an input corresponding to a point with small order, where the order divides the cofactor of the curve (see [Section 7](https://www.rfc-editor.org/info/rfc7748/#section-7)).

The Xwing draft makes no mention of this, and thus inherently embeds the MAY in its specification, The `x-wing` crate does not implement this check, and thus permits decapsulation of ciphertexts with non-contributory behaviour on the X25519 side.

However, this is incompatible with [Section 4.2 of `draft-irtf-cfrg-hybrid-kems-12`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hybrid-kems-12#section-4.2-4), which states:

> A scalar value of zero MUST NOT be used as a private key: it does not correspond to a meaningful Diffie-Hellman exchange and, for some curves such as P-256, is explicitly forbidden by the relevant standards and would yield a point-at-infinity encoding that departs from the fixed-size model used here. RandomScalar MUST NOT return a zero scalar; for the groups of interest this occurs only with negligible probability.

The Xwing defined using this draft (in [Section 4.2 of `draft-irtf-cfrg-concrete-hybrid-kems-04`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-concrete-hybrid-kems-04#name-mlkem768-x25519), which claims to be identical to `draft-connolly-cfrg-xwing-kem-10`) therefore *does* reject non-contributory behaviour (although it is currently inconsistent; [Section 3.1.2](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-concrete-hybrid-kems-04#name-curve25519-nominal-group) says that `RandomScalar` is the identity function for Curve25519, which violates the MUST NOT of `draft-irtf-cfrg-hybrid-kems-12`).

Now, for an honestly-generated key the chance of encountering this divergence is negligible:
- The Curve25519 scalar within a decapsulation key is a subslice of the output of `PRG(seed)` (via [`GenerateKeyPair`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hybrid-kems-12#name-key-generation) and [`expandDecapsKeyG`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hybrid-kems-12#name-using-a-nominal-group)).
- The ephemeral Curve25519 scalar used within encapsulation is the output of `RandomScalar(random(32))` (via [`prepareEncapsG`](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hybrid-kems-12#name-using-a-nominal-group)), which is constrained by the MUST NOT on the output of `RandomScalar` (once it is fixed for the concrete instantiation as I noted above).

However, for an adversarial ciphertext, there is nothing preventing the adversary from using zero as the ephemeral Curve25519 scalar, or a point of small order as the ephemeral public key. Indeed, I encountered this issue due to @FiloSottile exercising this specific edge case in age test vectors ([`hybrid_identity`](https://github.com/C2SP/CCTV/blob/main/age/internal/tests/hybrid_identity.go), [`hybrid_low_order`](https://github.com/C2SP/CCTV/blob/main/age/internal/tests/hybrid_low_order.go), [the test case generator](https://github.com/C2SP/CCTV/blob/1e3d2860d46e94e777e1b17c7a6f2436387e3ecc/age/internal/testkit/testkit.go#L198-L214)). So the probability of encountering this divergence is non-negligible, and we need to deal with it.

I'm going to discuss what to do about the drafts themselves with CFRG. However, as many people have already shipped Xwing, I expect many existing implementations will be allowing non-contributory behaviour, so there will likely be domains in which it is required that `x-wing` behave the way it currently does. So I'm thinking that the simplest fix to `x-wing` is:
- Ensure it never generates the zero scalar anywhere on encapsulation (which is allowed by the MAY in RFC 7748).
- Add a flag to the `DecapsulationKey` constructors (or alternate constructors) that enables the caller to select whether or not non-contributory behaviour is rejected at decapsulation time.

### PR DESCRIPTION
**Pull request by str4d**
Until the IETF RFC is published selecting a specific behaviour, we support both.

Closes RustCrypto/KEMs#364.

### CODE REVIEW

File: x-wing/src/lib.rs [hunk 1]
```diff
@@ -184,9 +191,47 @@ impl TryFrom<&[u8]> for EncapsulationKey {
...
+    /// behaviours: constructing a `DecapsulationKey` via [`KeyInit`] accepts
+    /// non-contributory behaviour for backwards-compatibility with existing usages, and
+    /// this method can be used to instead reject non-contributory behaviour. Once the RFC
+    /// is published, the default behaviour of `DecapsulationKey` will be altered to match
+    /// it.
+    ///
+    /// [RFC 7748]: https://www.rfc-editor.org/info/rfc7748/#section-5
+    /// [specifies]: https://www.rfc-editor.org/info/rfc7748/#section-6.1
+    /// [CFRG have decided]: https://mailarchive.ietf.org/arch/msg/cfrg/v9fEHQj3QTUpdu72AzjyyrY4j2g/
+    pub fn reject_x25519_non_contributory_behaviour(self) -> Self {
```
Comment (str4d) [inline x-wing/src/lib.rs]: I made this an extra method and didn't change the default behaviour, in order to make this a non-breaking change. But then I remembered that changing from `kem::Decapsulate` to `kem::TryDecapsulate` is breaking anyway, so preserving behaviour is not technically necessary as long as we provide a way to reach either state.

**Comment by tarcieri** [comment]
My personal inclination would just be to always reject, particularly if we switch to implementing `TryDecapsulate`. That matches the current behavior of `dhkem`.

I would also suggest against adding new methods to the public API, at least if their utility would be eliminated by a future change to the RFC.

**Comment by rozbb** [comment]
My 2c: it is odd to me to make `TryDecapsulate` have a specific behavior not stated in the spec being implemented. One way to avoid this while preserving the API is to just have another type, like `DecapsulationKeyRejectNonContrib`, which impls `TryDecapsulate`

**Comment by tarcieri** [comment]
@rozbb that's an interesting direction and could be done in a purely additive manner, then when the RFC is clarified we can remove one and keep the other

**Comment by str4d** [comment]
[Force-pushed](https://github.com/RustCrypto/KEMs/compare/faabbc960470b38a51f4aec07ce7f1d1c381d401..f5dd74e238bfebc0b36754923ccc6157d75641e6) to make this a non-breaking change via adding `DecapsulationKeyRejectNonContrib`.

**Review by rozbb — APPROVED** [review]

File: x-wing/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+use core::fmt;
+
+/// Error type.
+#[derive(Clone, Copy, Debug)]
+#[non_exhaustive]
+pub enum Error {
+    /// Decapsulation failed.
+    Decapsulation,
+}
```
Comment (tarcieri) [inline x-wing/src/error.rs]: I'm wondering if this type being an `enum` is YAGNI. Perhaps it could just be `pub struct DecapsulationError` or `pub struct Error`.

Comment (tarcieri) [reply x-wing/src/error.rs]: I was trying to think of other cases and looked at `dhkem`, which also has a `Length` error variant: https://github.com/RustCrypto/KEMs/blob/master/dhkem/src/error.rs#L3-L12

Comment (rozbb) [reply x-wing/src/error.rs]: I agree it doesn’t need to be an enum. It should be particular to `DecapsulationKeyRejectNonContrib` though, since we could reasonably add another decap type later. So either

1. `pub struct Error` inside of a `mod rejectnoncontrib`. Or
2. `pub struct NonContribError`

Comment (tarcieri) [reply x-wing/src/error.rs]: `pub struct DecapsulationError`  might be fine as it definitionally can't apply to something with infallible decapsulation

**Comment by tarcieri** [comment]
Three ways to land this:

1. Switch the error type to a unit `struct`
2. Add another variant to it to justify the `enum`
3. Merge this as-is, but with a followup issue that's a release blocker

**Comment by rozbb** [comment]
I vote unit struct. We can have a breaking upgrade in the future imo. The feature is really unlikely to be used outside of `rust-hpke` I reckon

**Comment by str4d** [comment]
[Force-pushed](https://github.com/RustCrypto/KEMs/compare/f5dd74e238bfebc0b36754923ccc6157d75641e6..55bd3fee6e94bebd9e78a2d15b970b55fb5740f4) to switch to a zero-sized struct. If you wanted `pub struct DecapsulationError(())` instead (I don't think that's what you meant), let me know.

**Review by tarcieri — APPROVED** [review]
